Introduce the ability to run FIPS KAT tests on demand

simo5 opened the pull request:
This patch affects only the FIPS module.

This is an attempt to devise a method to defer KAT tests to the first time an algorithm
is required to be actually used.

The on_demand facility run a single test identified by type of KAT and name, and reuses the existing self-test infrastructure to run the test.

Until the on demand test is marked as passed an attempt to use the algorithm implementation will always end up calling FIPS_kat_on_demand().

This function takes a lock to prevent multiple threads from attempting to execute in parallel, it then proceeds to execute the test.
A special thread only key is used to detect if the thread is engaged in the self test, if the thread loops into calling  FIPS_kat_on_demand() this is detected and the function returns a -1 state.
When this happens the code keeps executing without locally marking the test as passed.

Once the test execution is complete the lock is released, and other threads can proceed.

The SLH-DSA algorithm is switched to be executed on demand. And the SLH test implicitly test this code, so no additional test are provided.

Fixes #27783

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -1238,3 +1259,128 @@ void OSSL_INDICATOR_get_callback(OSSL_LIB_CTX *libctx,
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_signature_tests[] = {
+    { "SLH-DSA-SHAKE-128f", 0 },
```
With deferred testing, it seems reasonable to include `SLH-DSA-SHAKE-128s` as well.  This is recommended by the standards but not mandated.

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -97,6 +99,9 @@ typedef struct fips_global_st {
 
+    /* Guards access to ondemand self-test */
+    CRYPTO_RWLOCK *ondemand_lock;
+    CRYPTO_THREAD_LOCAL ondemand_key;
```
Do we want this per libctx?  That seems a little excessive.
It's also not consistent with the single global flag for the one deferred algorithm.

simo5 replied:
I tried to have them as global static initially but all the infrastructure around RUN_ONCE() prevents their usage in FIPS_MODULE, so for now I stored them on the context, if we think we can relax the code that blocks those initializers for the fips provider this stuff can be very easily moved off again.

nhorman replied:
you can use the CRYPTO_THREAD_run_once function directly in the FIPS module, rather than the RUN_ONCE macros.  I had to do the same thing for the CRYPTO_THREAD_[set|get]_local_ex apis.

simo5 replied:
Alright, if this is allowed I can change it that way!

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -1238,3 +1259,128 @@ void OSSL_INDICATOR_get_callback(OSSL_LIB_CTX *libctx,
+    int idx = 0;
+    int ret = 0;
+
+    if ((fgbl = ossl_lib_ctx_get_data(libctx, OSSL_LIB_CTX_FIPS_PROV_INDEX)) == NULL)
```
Note that this is an expensive operation.  It is not a problem given them way SLH-DSA uses a global flag but it could become one if other algorithms get naïvely converted.

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -1238,3 +1259,128 @@ void OSSL_INDICATOR_get_callback(OSSL_LIB_CTX *libctx,
+    if (ondemand_tests_status[type][idx].pass)
+        return 1;
+
+    /*
+     * now that we found the test, before we do anything else make sure
+     * a local test is not already in progress or we'll deadlock
+     */
+    local_key = CRYPTO_THREAD_get_local(&fgbl->ondemand_key);
+    if (local_key != NULL) {
+        /* self test occurring in this thread, let it continue */
+        return -1;
+    }
+
+    if (CRYPTO_THREAD_write_lock(fgbl->ondemand_lock)) {
```
Why have both a thread local indicating that the test is running and a global lock to run the test under?

I don't see how the thread local helps.  If we're running the test, we're running the test and won't be interrupted until it's finished.  If we're not running it, we'll never run it.

simo5 replied:
The tests will call back in to the functions that called the test right away, so we need a way to make those functions go through and not try to run self tests again in a loop while we are testing. But we do not want any other thread to be allowed to just execute the same functions until the self-test is complete.
So the lock blocks all other threads, while the thread-local storage allow us to avoid recursively taking the lock and deadlocking on the thread that is currently running the test.

paulidale reviewed (commented):
A very good start.

slontis commented on providers/implementations/signature/slh_dsa_sig.c.in:
```diff
@@ -39,6 +41,70 @@ static OSSL_FUNC_signature_dupctx_fn slh_dsa_dupctx;
+        case -1:
+            /*
+             * self test in progress for this thread so we let operations
+             * continue without changing any other state
```
You cannot use the algorithm before the self test has finished, so wouldnt it have to wait here?

slontis commented on providers/implementations/signature/slh_dsa_sig.c.in:
```diff
@@ -153,7 +219,7 @@ static int slh_dsa_signverify_msg_init(void *vctx, void *vkey,
     SLH_DSA_KEY *key = vkey;
 
-    if (!ossl_prov_is_running()
+    if (!slh_dsa_self_check(ossl_slh_dsa_key_get0_libctx(key))
```
Why are there multiple calls to the self_check here.. The object is setup by the newctx(). That should be the point at which the self_check runs if it need to.

simo5 replied:
I backed out some of these, specifically I had to back out the checks on keymgmt gen_init because pairwise_fail_test fail test assumes the init always work and the fail is on the keygen ...

slontis reviewed (commented):
There is already an OSSL_PROVIDER_self_test() call that should be able to run all of the self tests on demand (including SLH_DSA). So the logic where you skip on_demand tests is not quite right. It needs to skip on power on, but not during manual on_demand.

nhorman commented on providers/implementations/signature/slh_dsa_sig.c.in:
```diff
@@ -39,6 +41,70 @@ static OSSL_FUNC_signature_dupctx_fn slh_dsa_dupctx;
+     * until this lock is initialized we do not want to recurse into
+     * the self testing infrastructure
+     */
+    if (!slh_dsa_sig_kat_pass) {
```
This is a global variable that is both read and written without the benefit of a lock protecting it.  I don't think we have a test to exercise multiple threads calling into the FIPS provider for the same alg, so I'm not sure we  trigger it, but I would imagine that this has the potential to trigger thread sanitizer RAW/WAR errors.

The same appears true for all the *_kat_pass variables

simo5 replied:
I was trying to use CRYPTO_atomic_load_int at some point, but those APIs are somewhat limited and also artificially require to pass in a CRYPTO_RWLOCK object because some odd platform doesn't have atomic ?
Locking here would be expensive, an atomic read is all we should really ever need.

That said, even without atomics the worst case would be that the variable is read twice in two threads and the tests invoked twice, but because of the way tests work, only one thread will execute them while the other is locked. At the end of the test, both threads will write 1 in the variable, and that's the end of the story.

So for the submission I decided to avoid even the atomics because I am not sure that we actually ever really need to have this strict synchronization between threads, given FIPS_kat_on_demand() is the actual synchronization point that guarantees proper behavior.

nhorman replied:
I agree with the assessment that the worst case scenario is that we run additional self tests in a multithreaded environment, but in that case, if you want to avoid the atomics, we should likely annotate the usage to avoid getting false positive failures in the field.  Something like TSAN_FAKE_LOCK from crypto/threads_pthread should likely do it.

nhorman replied:
Well, as I noted, I don't know if it will complain, and it definitely wont in our testing right now, because we don't have a test that attempts to do multiple self tests from multiple threads in parallel, but its strictly speaking a race condition, as there is no ordering on this, and tsan tends to be pretty bull-headed on that, as it doesn't really take arch memory nuances into account.  If you don't want to add the annotations, I'd be ok with a comment noting that we're comfortable with the potential for multiple repeated self tests, as the consequence of avoiding the atomics/locking seems like a reasonable trade off.

simo5 commented on providers/fips/fipsprov.c:
```diff
@@ -1238,3 +1259,128 @@ void OSSL_INDICATOR_get_callback(OSSL_LIB_CTX *libctx,
+         * thread, does not matter what value, as long as
+         * it is not NULL, Cool?
+         */
+        if (!CRYPTO_THREAD_set_local(&fgbl->ondemand_key, (void *)0xC001))
```
Yeah but I thought the per-thread part was done by the master_key (which *is* a thread local variable so there is a different master_key per thread) which implies an array indexed by id for each thread, so I do not fully get why then you have a table stored in the id slot instead of just storing the data directly, no other thread should be able to access that same table, so my reading is that you have a table where you ever only store one entry? Or am I reading it all wrong ?

simo5 commented:
> > There is already an OSSL_PROVIDER_self_test() call that should be able to run all of the self tests on demand (including SLH_DSA). So the logic where you skip on_demand tests is not quite right. It needs to skip on power on, but not during manual on_demand.
> 
> Yes I realized that, but decided to defer handling that until we are ok with the general direction of these patches.
> 
> I consider that a problem that would block accepting this PR until resolved.
> 
> That said one of the reasons I did not address that is that once test are on-demand per algorithm perhaps we should also require that applications are able to expresee what they want to test on demand? Maybe with a way to express "all and everything under the sun", but I am not sure we should just run all tests implicitly otherwise?
> 
> ^ I am totally flexible on this point , would like to discuss what people should expect here.

FYI I changed SELF_TEST_post() and SELF_TEST_kats() to propagate the information about whether tests are being requested so that on demand tests will also be tested when requested via ossl_provider_self_test(), I pushed it as a separate commit message, and the whole PR has been rebased to include changes done according to the comments from the various reviwers

slontis commented:
I would probably change the terminology so the struct variable is not 'ondemand' as there seem to be 2 different concepts here
i.e.
1) Run all the tests manually on demand
2) Delay running an algorithm self test until first use (i.e. This is a lazy init).

'deferred' maybe?

beldmit commented:
Why not call it just lazy then? :)

simo5 commented on providers/fips/fipsprov.c:
```diff
@@ -1238,3 +1246,224 @@ void OSSL_INDICATOR_get_callback(OSSL_LIB_CTX *libctx,
+         * thread, does not matter what value, as long as
+         * it is not NULL, Cool?
+         */
+        if (!CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_FIPS_DEFERRED_KEY,
```
Please read the other (resolved) comments in this PR, I already explained it there.

simo5 commented on providers/implementations/signature/slh_dsa_sig.c.in:
```diff
@@ -209,7 +231,7 @@ static int slh_dsa_sign(void *vctx, unsigned char *sig, size_t *siglen,
     size_t n = 0;
 
-    if (!ossl_prov_is_running())
+    if (!slh_dsa_self_check(ctx->libctx))
```
I do not believe there is any problem with running tests concurrently, it won't happen though (via the deferred mechanism) because we allow a single deferred test of any kind to be run at any given time, that is guaranteed by the deferred_lock.

simo5 commented:
> Really impressive work here... Good job @simo5 !!
> 
> This seems like a behavior change that developers should be aware of. Do you think we should add some documentation about the on-demand testing approach, or is the existing FIPS documentation sufficient?

I do not know, I'll defere to other openssl developers, perhaps a changelog item will be sufficient.

jvdsn commented on providers/implementations/signature/slh_dsa_sig.c.in:
```diff
@@ -39,6 +41,26 @@ static OSSL_FUNC_signature_dupctx_fn slh_dsa_dupctx;
+};
+#endif
+
+static int slh_dsa_self_check(OSSL_LIB_CTX *libctx)
```
Would it make sense to split this up in two functions? Key generation and signature generation / verification are different operations. Should a caller requesting SLH-DSA signature verification have to pay for the key generation KAT? Or even for the signature generation KAT?

EDIT: actually, never mind, I'm confused why key generation is here at all. The key generation is implemented in kmgmt? The (lesser) point still stands about whether or not signature verification should trigger a signature generation KAT.

simo5 replied:
Currently we run 3 tests in totoal for SLH-DSA:
1. SHA2-based keygen
2. SHA2-base signature
3. SHAKE base signature

These tests are *slow*

While I was reviewing the keygen tests I realized that they include a full signature and verification loop after the key because:
keygen test means calling slh_dsa_gen()
slh_dsa_gen() calls slh_dsa_fips140_pairwise_test()
slh_dsa_fips140_pairwise_test() calls ossl_slh_dsa_sign() and ossl_slh_dsa_verify()

Given the inputs of the keygen test are static, then these internal sig/ver calls are effectively a KAT for the SLH-DSA-SHA2-128f algorithm.

Therefore by calling the keygen test we effectively already cover both (1) and (2) and calling (2) again would be redundant.

So I call only (1) also in the signature function so that in total we ever only perform (1) once regardless of which pf the two facilities (keygen or sigver) is called first and it covers both.

The end result is that in total we only ever do 2 tests and cover 3 scenarios instead of calling 3 tests for the same 3 scenarios.

simo5 replied:
I now pushed a commit that removes the SLH-DSA-SHA2-128f sigver test with a justification in the commit message

simo5 replied:
If you think the internal pairwise test does not behave as a sufficient KAT please let me know and I will try to fix it so that it fully covers that role, reducing the cost of these tests is somewhat important to me.

jvdsn replied:
Okay, I understand what you're trying to do now. However, a PCT is not sufficient to meet the CAST requirements. This was the case in FIPS 140-2, but in FIPS 140-3 there must be a comparison of the generated signature. Specifically (FIPS 140-3 IG 10.3.A):

>The CAST shall use the SLH-DSA signature generation algorithm (i.e., Algorithm 22 in FIPS 205) or SLH-DSA signature generation internal algorithm (i.e., Algorithm 19 in FIPS 205), and for a KAT, using fixed/predetermined values (i.e., for Algorithm 22: SK, M, addrnd if non-deterministic variant, and ctx which may be an empty string; for Algorithm 19: SK, M’ and addrnd if non-deterministic variant) to **compare the resulting outputs to the pre-computed value (i.e., SIG).**

> The CAST shall use the SLH-DSA signature verification algorithm (i.e., Algorithm 24 in FIPS 205) or SLH-DSA signature verification internal algorithm (i.e., Algorithm 20 in FIPS 205), and for a KAT, using fixed/predetermined values (i.e., for Algorithm 24: PK, M, SIG, and ctx which may be an empty string; for Algorithm 20: PK, M’, SIG) to verify the signature (i.e., returns true).

You can combine the SigGen and SigVer CASTs, sure, if you perform a comparison of the generated signature in between. But that's not the same as a PCT.

simo5 replied:
So all I need is to compute the signature, save it in the test vector and compare the signature while doing the PCT and that will double as a KAT, right ?

jvdsn replied:
@simo5 well, the problem is that a KAT requires deterministic inputs. So in the very specific scenario of the keygen KAT, this would work. But the PCT is also used in other contexts, when generating key pairs. If you were to introduce a comparison at that stage, it would fail, since the key pair would be different.

Perhaps you can hack it in such a way that the KAT inside the PCT is only performed when the higher-level keygen KAT is performed. It's a little bit un-intuitive.

simo5 replied:
It might be easier to show what I mean with an optional patch

jvdsn replied:
I think this would be a valid approach, however I am wondering if it wouldn't just be easier to omit the PCT when you're performing KATs? Then you can have a separate explicit KAT for SigGen / SigVer without the performance penalty.

simo5 replied:
Ok I will look into removing this part and simply skip the pairwise test, re-instating two simple sig tests in the signature path

jvdsn commented on include/internal/fips.h:
```diff
@@ -16,6 +16,28 @@
 /* Return 1 if the FIPS self tests are running and 0 otherwise */
 int ossl_fips_self_testing(void);
 
+/* Ondemand KAT tests methods */
```
In a couple places the comments still refer to on demand rather than deferred.

jvdsn reviewed (commented):
I think this is a reasonable start. One general pitfall with this approach is multithreading, but it seems like this is being dealt with. I'm not an expert on that, so I'll assume you implemented it correctly.

There's many algorithm-specific nuances as well. Those can be discussed as the PR progresses. Here's one for free that I've seen other FIPS modules trip on: before performing AES-CTR _decryption_, you must perform a KAT that tests the _forward_ AES cipher function.

simo5 commented:
Given a test flaked I addressed the remaining places where `on demand` was being used instead of `deferred` including commit messages, and pushed a rebase so all tests are re-run.

jvdsn commented on providers/implementations/keymgmt/slh_dsa_kmgmt.c.in:
```diff
@@ -339,6 +365,13 @@ static int slh_dsa_fips140_pairwise_test(const SLH_DSA_KEY *key,
                           sig, &sig_len, sig_len) != 1)
         goto err;
 
```
OpenSSL uses `OSSL_SELF_TEST_oncorrupt_byte(st, sig);` to corrupt self-tests, so if this function were to do both PCT and KATs, it would have to have at least 3 of those I think: one to corrupt the PCT, and two to corrupt the SigGen and SigVer KATs.

simo5 replied:
Is this a requirement for self-tests?

slontis replied:
Its a requirement for testing. This sounds like it is getting complicated.

jvdsn commented on providers/implementations/keymgmt/slh_dsa_kmgmt.c.in:
```diff
@@ -373,7 +406,7 @@ static void *slh_dsa_gen(void *genctx, const char *alg)
 #ifdef FIPS_MODULE
-    if (!slh_dsa_fips140_pairwise_test(key, ctx)) {
+    if (!slh_dsa_fips140_pairwise_test(key, gctx, ctx)) {
         ossl_set_error_state(OSSL_SELF_TEST_TYPE_PCT);
```
I'm wondering if the error state should change depending on whether or not the kat_sig parameter is present. If it is present, then you're doing a KAT, so that would be the appropriate error state. If it's not present, you're doing a PCT.

simo5 replied:
Well I am doing both, but it I do not know if it matters in terms of raising the error.

simo5 commented on providers/implementations/keymgmt/slh_dsa_kmgmt.c.in:
```diff
@@ -339,6 +365,13 @@ static int slh_dsa_fips140_pairwise_test(const SLH_DSA_KEY *key,
+    /* KAT test for optimized KEYGEN + Signature KAT */
+    if (gctx->kat_sig != NULL && gctx->kat_sig_len != 0) {
+        if ((sig_len != gctx->kat_sig_len)
+            && (memcmp(gctx->kat_sig, sig, sig_len) != 0))
```
yeah, but I can probably drop checking for gctx->kat_sig_len != 0 given that we immediately after check if sig_len != gctx->kat_sig_len

simo5 commented on providers/implementations/keymgmt/slh_dsa_kmgmt.c.in:
```diff
@@ -433,6 +476,9 @@ static void slh_dsa_gen_cleanup(void *genctx)
         return;
 
+#ifdef FIPS_MODULE
+    OPENSSL_clear_free(gctx->kat_sig, gctx->kat_sig_len);
```
I can turn this into a simple OPENSSL_free(), np

simo5 commented:
> > This seems like a behavior change that developers should be aware of. Do you think we should add some documentation about the on-demand testing approach, or is the existing FIPS documentation sufficient?
> 
> Lets not try to make this general purpose. It is intended for slow algorithms, not things that run fast, which is most things.

My plan is to allow downstream certification to easily decide to make as many algorithms on-demand as they want with minimal patching, and perhaps no patching at all eventually, but just build time configuration.

While many algorithms have relatively fast tests we've always had customer's complaints that in FIPS mode systems have a too high startup penalty. And in so many cases (especially for small binaries) 90% of the tests are just wasted heat (think of the sha256sum binary which only uses sha256 and nothing else but currently has to pay the price of testing all supported asymmetric operation as well as symmetric ciphers).

So I really want to make it possible for this to be a general purpose mechanism so we can fix the general problem at once. SLH-DSA was only the drop that made the bucket overflow, but it is not the only problem.

simo5 commented:
I rebased this work to avoid conflating keygen and signature tests, now the keygen test simply skips the pairwise consistency check when the code is performing general self tests or a deferred self test.

I also renamed some of the variables to make it clear this is FIPS-only machinery and move the state for tests to use defines so that there are less "magic" numbers being used. These were basically just cosmetic changes to improve naming around. The only effective change compared to previous code is that now we record at the implementation level that the test is in progress so that the code can skip things like pairwise tests as necessary.

simo5 commented on providers/fips/self_test.c:
```diff
@@ -381,7 +381,7 @@ int SELF_TEST_post(SELF_TEST_POST_PARAMS *st, int on_demand_test)
     }
 
-    if (!SELF_TEST_kats(ev, st->libctx)) {
+    if (!SELF_TEST_kats(ev, st->libctx, on_demand_test)) {
```
It is already documented:
"when do_deferred is 1 also run deferred tests"

simo5 replied:
well the on_demand tests running all test was an explicit request during review to maintain the "on_demand" part consistent, so the change I made here is to actually to conform to the current documentation ?
If we did not do this we'd have to provide a new public interface to be able to run individual tests on demand, and I do not think we want that. When tests are run on demand it is perfectly fine to run them all IMHO.

simo5 commented on providers/fips/fipsprov.c:
```diff
@@ -1266,85 +1266,23 @@ static void deferred_deinit(void)
     /*
-     * now that we found the test, before we do anything else make sure
-     * a local test is not already in progress or we'll deadlock
+     * before we do anything, make sure * a local test is not already in
```
NOTE: I already fixed this stray asterisk locally and will push it after the next round of reviews with any requested changes

beldmit commented:
Ping @openssl/committers for review

paulidale commented on providers/fips/fipsprov.c:
```diff
@@ -1238,3 +1246,163 @@ void OSSL_INDICATOR_get_callback(OSSL_LIB_CTX *libctx,
+            state = FIPS_kat_deferred(libctx, &tests[i]);
+            switch (state) {
+            case FIPS_DEFERRED_TEST_IN_PROGRESS:
+                /* self test in progress for this thread so we stop here. */
```
I think this sums up the situation better:
```suggestion
                /* self tests in progress so we assume they pass and return success. */
```

However, I'm not sure this is correct behaviour.  Blocking until the self tests either pass or fail seems more reasonable.  Doing that would remove this state I suspect.

simo5 replied:
A comment about locking is above the for loop, do you mean an additional comment to also explain that we let one thread go to complete the ongoing tests ?
I thought this comment would hint at that but I welcome clearer language.

paulidale replied:
How about:
```
                /*
                 * The self tests are currently running.  If we are the thread that is running them
                 * then the algorithm is allowed so return okay.  Without this, deadlock will occur.
                 */
```

simo5 replied:
we raced on it, let me know if the fixup I just push is good enough

paulidale replied:
That works.

simo5 commented:
> I've doubts about wanting to defer all self tests or even allowing per algorithm deferrals. The inability to swap the DRBG limits the algorithms that can be self tested, so we're never going to have total control over what can and cannot be deferred.

This is a very important feature for us, runtime KAT tests cost a lot in terms of process startup, to the point we had to disable SLH-DSA in our current build because the tests are so onerous they were slowing everything down too much.

Most tests that (ab)use the regular DRBG can probably be changed to use fixed values and not pull for the DRBG at all. But even if some tests still need to be run early for now that is probably acceptable.

> I've also got concerns about the threading and locking.

This is fully tested and carefully written, but we can definitely tweak as needed.

paulidale commented on include/internal/fips.h:
```diff
@@ -16,6 +16,33 @@
 int ossl_fips_self_testing(void);
 
+/* Deferred KAT tests categories */
+#  define FIPS_DEFERRED_KAT_INTEGRITY 0
```
The digest/mac used for the integrity check falls into the same category but the granularity for this isn't present.

simo5 replied:
so, the thing is that ideally we change the code so that each algorithm fully expresses all the dependencies, and then we change the code to always go through the "on demand" testing, even for the initial tests so that which test has been done (and passed) is recorded in the individual provider structures.

Then by configuration we'd define which tests really are deferred and which are not.

This would make it impossible to make mistakes because a function that requires others to be tested would always trigger their tests first (unless they have already passed).

I did not want to go that far right away in this PR, I think that would be appropriate for a follow up PR once the deferral mechanism is accepted.

paulidale commented on include/internal/fips.h:
```diff
@@ -16,6 +16,39 @@
+#  define FIPS_DEFERRED_KAT_SIGNATURE 6
+#  define FIPS_DEFERRED_KAT_KDF 7
+#  define FIPS_DEFERRED_KAT_KA 8
+#  define FIPS_DEFERRED_MAX 9
```
MAC & RAND are missing.
MAC aren't tested independently but RANDs (drbgs) are.

paulidale replied:
I was more thinking of someone who uses CMAC but not the DRBG.
Currently, the only CMAC test we have is the DRBG one which is lots more expensive than just testing CMAC.  Having the CMAC trigger this is possible but it isn't self contained.

However, we don't want to add an individual CMAC test that runs unconditionally since the DRBG can do it for us.  Your _also satisfies_ would deal with this case I think.

paulidale commented on providers/implementations/keymgmt/slh_dsa_kmgmt.c:
```diff
@@ -115,7 +138,7 @@ static int slh_dsa_import(void *keydata, int selection, const OSSL_PARAM params[
     int include_priv;
     struct slh_dsa_import_st p;
 
-    if (!ossl_prov_is_running()
```
I don't think this and later changes to replace `ossl_prov_is_running()` are required.  It's not possible to call this and later functions without having first called `slh_dsa_new_key()` which includes the self check call.

The same applies to `slh_dsa_sig.c` where only `slh_dsa_newctx()` needs the self test call but the others don't.

paulidale commented:
This looks very close now.

paulidale commented on providers/fips/self_test_kats.c:
```diff
@@ -1237,22 +1266,181 @@ int SELF_TEST_kats(OSSL_SELF_TEST *st, OSSL_LIB_CTX *libctx)
+        return 0;
+
+    /* self_test_digests */
+    if (type == FIPS_DEFERRED_KAT_DIGEST) {
```
It's probably worthwhile changing this to a switch statement.

simo5 replied:
ok

paulidale commented:
Great work.  Just a couple of nits really.

simo5 commented:
I rebase on lastest master and I think all the requested changes are in now.

slontis reviewed (commented):
This is not for this PR, but if we are going down the path of deferring for most algorithms I would like to see some form of design doc for it.
There are quite a few things to consider, such as

1) Many algorithms have multiple sub algorithms, and some of these sub algorithms are choices. For many algorithms this choice is selected late.
e.g.
- SSKDF hash 3 modes HMAC, KMAC or HASH.
- PBKDF2 uses HMAC with any FIPS digest (except for Shake)

2) Because nothing is currently deferred there are some assumptions being made about higher level functions that contain sub-algorithms that dont need a test..

e.g.
- gmac doesnt have a self test since it is covered by aeg-gcm running.
- kmac uses keccak (sha3 self test)
- HMAC and SHA256 are done by the integrity self test.

Maybe deferring digests is not a good idea, due the large numbers of places they are used.

3) The dependancy chain of every single algorithm needs to be carefully considered.
Quite a few depend on random (DRBG).

4) Hopefully there are no cyclic depedencies but I would not be suprised if there are.

5) The DRBG self tests were setup to not drain entropy, maybe that is not such an issue if they are deferred?

slontis commented on providers/implementations/keymgmt/slh_dsa_kmgmt.c:
```diff
@@ -294,7 +318,8 @@ static int slh_dsa_fips140_pairwise_test(const SLH_DSA_KEY *key,
     /* During self test, it is a waste to do this test */
-    if (ossl_fips_self_testing())
+    if (ossl_fips_self_testing()
+        || slh_key_gen_deferred_tests[0].state == FIPS_DEFERRED_TEST_IN_PROGRESS)
```
Its not likely but potentially this could be in a failed state?

slontis commented on providers/fips/self_test_kats.c:
```diff
@@ -1237,22 +1266,179 @@ int SELF_TEST_kats(OSSL_SELF_TEST *st, OSSL_LIB_CTX *libctx)
+ */
+int SELF_TEST_kats_single(OSSL_SELF_TEST *st, OSSL_LIB_CTX *libctx,
+                          int type, const char *alg_name)
+{
```
Would it be cleaner to modify all the self tests loops rather than do this by passing the alg as an optional parameter?

e.g.
````
static int self_test_digests(OSSL_SELF_TEST *s,..., const char *alg)
{      
}
````

This then keeps the logic in opne place rather than being duplicated across functions.

simo5 commented on include/internal/fips.h:
```diff
@@ -16,6 +16,42 @@
+ * really be deferred. Keep it commented here as a reminder.
+ * #  define FIPS_DEFERRED_KAT_INTEGRITY 0
+ */
+#  define FIPS_DEFERRED_KAT_CIPHER 1
```
Yes there is a reason.
Ids can be easily used in switch statements and do not have as much of a performance drawback, can be easily changed, can be checked (and are I got a failure for that during development) for incorrect duplication in switch statements by compilers, and can be ordered naturally.
There is nothing good about using strings unless you have to provide ABI compatibility for external applications and this is a completely internal aspect of the provider, never exposed externally.

paulidale replied:
I almost suggested using an enum here instead of defines because it is internal.

paulidale replied:
We cannot use enums for public APIs but they are fair game for internal ones.  That's part of the reason for their scarce usage.  The rest is historical.

I agree about not holding this up needlessly.  Maybe make the change if the second reviewer finds something else.

simo5 commented:
> This is not for this PR, but if we are going down the path of deferring for most algorithms I would like to see some form of design doc for it.

Yes, if we are going to overhaul this in that direction a design document makes sense.

> There are quite a few things to consider, such as
> 
>     1. Many algorithms have multiple sub algorithms, and some of these sub algorithms are choices. For many algorithms this choice is selected late.
>        e.g.

See my comments in this huge PR about "also_satisfies" for how we can handle some of these things.

>     * SSKDF hash 3 modes HMAC, KMAC or HASH.

Generally you need to test a mechanism with only one of them, we could have some mechanism to declare equivalency, but given HMAC is going to be the most used MAC for years to come, it will be simpler to just use HMAC here. At some point pragmatic choices trump theoretical benefits and complicated architectures to handle corner cases :)

>     * PBKDF2 uses HMAC with any FIPS digest (except for Shake)

I am pretty sure HMAC only need to be tested with one Hash, again we could devise an equivalency method, but what for? If the and underlying Hash has already been explicitly tested we'd still have to run the HMAC KAT for the HMAC part, and unless you want to provide a choice of different KATs for HMAC you have already chosen the underlying testing algorithm when the test was written.

Note that is the HMAC KAT uses SHA-256, and then the user tries to use HMAC with SHA-512 then at the moment the HMAC function tries to initialize the digest it would trigger that digest self test check, and if the hash had not been KAT tested it would trigger the SHA-512 KAT before allowing the HMAC initialization to complete.

There is no real complexity here provided we correctly identify entry points for any of the function to trigger the tests.

>     2. Because nothing is currently deferred there are some assumptions being made about higher level functions that contain sub-algorithms that dont need a test..

Yes see my comment about "also_satisfies"

> 
> 
> e.g.
> 
>     * gmac doesnt have a self test since it is covered by aeg-gcm running.

GMAC can't be used for anything but AES-GCM in FIPS mode, so this is probably not a good example.

>     * kmac uses keccak (sha3 self test)

Technically it uses cSHAKE, but then again we just need to properly mark self-test entry points if we want to defer hashing algorithms self-tests.

>     * HMAC and SHA256 are done by the integrity self test.

And they would be marked as passed there (see also_satisfies proposal in my comments)

> Maybe deferring digests is not a good idea, due the large numbers of places they are used.

We do not have to defer absolutely everything, but I think it would be useful to defer as much as possible in many cases.

>     3. The dependancy chain of every single algorithm needs to be carefully considered.
>        Quite a few depend on random (DRBG).

Yes whenever an algorithm is converted to deferred the dependency chain needs to be evaluated and all dependents need to grow a dependency.

>     4. Hopefully there are no cyclic depedencies but I would not be suprised if there are.

Cyclic dependencies would not be an issue, during a test no algorithm usage is blocked, and all of the algorithms being used can be marked as passed. The "also_satisfies" method is one way to do it, another way is to collect all algorithms being run on the "testing thread" in a queue in FIPS_kat_deferred() (where we `return FIPS_DEFERRED_TEST_IN_PROGRESS;`) and then mark them all as passed in one swoop.

>     5. The DRBG self tests were setup to not drain entropy, maybe that is not such an issue if they are deferred?

See Pauli's response on this one.

simo5 commented:
> Btw the also_satisfies trick is only useful when a test is equivalent, if there is an actual dependency where FIPS mandates independent testing of the algorithm the current algorithm depends on we'll grow an actual "depends" list attached to FIPS_DEFERRED_TEST, which will cause the other test to be run first (potentially recursively a few times) before the test for the requested algorithm is run.

Thinking about this a little more the dependencies will largely be implicit anyway, by simply calling the function a higher level algorithm depends on it should trigger self-tests, so an explicit "dependency" chain would only be needed for odd cases where the actual test we want to run is not implicitly depended on. For example one could decided that the HMAC test becomes a "dependency" for the SHA-256 test so that by running the HMAC test which "also_satisfies" the SHA-256 test we run the actual SHA-256 test only once and as a bonus we get also HMAC passed.

Note that this is just speculation and we should really move any further discussion to a design document as it has been proposed.

openssl-machine commented:
This pull request is ready to merge

beldmit commented:
Merged. Many thanks for the contribution!
